Remove some `ref` patterns from the compiler

WaffleLapkin opened the pull request:
Previous PR: https://github.com/rust-lang/rust/pull/105368

r? @Nilstrieb

WaffleLapkin committed:
`rustc_const_eval`: remove `ref` patterns (+some pattern matching imps)

WaffleLapkin committed:
`rustc_data_structures`: remove `ref` patterns and other artifacts of the past

WaffleLapkin committed:
`rustc_expand`: remove `ref` patterns

Nilstrieb commented on compiler/rustc_const_eval/src/interpret/operand.rs:
```diff
@@ -533,11 +533,11 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
+        let op = match mir_op {
             // FIXME: do some more logic on `move` to invalidate the old location
-            Copy(place) | Move(place) => self.eval_place_to_op(place, layout)?,
+            &(Copy(place) | Move(place)) => self.eval_place_to_op(place, layout)?,
```
Not sure if this is an improvement here, maybe
```suggestion
            Copy(place) | Move(place) => self.eval_place_to_op(*place, *layout)?,
```
would be even nicer?

WaffleLapkin replied:
I don't think derefing here is nicer. Another alternative is
```suggestion
            &Copy(place) | &Move(place) => self.eval_place_to_op(place, layout)?,
```

Nilstrieb commented on compiler/rustc_const_eval/src/transform/promote_consts.rs:
```diff
@@ -849,54 +849,50 @@ impl<'a, 'tcx> Promoter<'a, 'tcx> {
-                        tcx.lifetimes.re_erased,
-                        ty::TypeAndMut { ty, mutbl: borrow_kind.to_mutbl_lossy() },
-                    );
+            let StatementKind::Assign(box (_, Rvalue::Ref(region, borrow_kind, place)))= &mut statement.kind else {
```
```suggestion
            let StatementKind::Assign(box (_, Rvalue::Ref(region, borrow_kind, place))) = &mut statement.kind else {
```

Noratrieb reviewed (commented):
Thanks, most of those look way better now! One `let_else` fomatting fix (really excited for the day when rustfmt will finally work with it :D) and a nit that you may or may not address, whatever you prefer.
r=me then

RalfJung commented on compiler/rustc_const_eval/src/const_eval/eval_queries.rs:
```diff
@@ -167,7 +167,7 @@ pub(super) fn op_to_const<'tcx>(
     };
     match immediate {
-        Left(ref mplace) => to_const_value(mplace),
+        Left(mplace) => to_const_value(&mplace),
```
Doesn't this change behavior? Previously we took a reference to the existing `immediate` local, now we move `mplace` to a new location and take a reference to that. Why is that better?

WaffleLapkin replied:
I thought that LLVM is smart enough to remove unnecessary moves, but it seems [it is not](https://godbolt.org/z/exWTznxos). I'll revert this change 😓

kadiwa4 commented on compiler/rustc_const_eval/src/interpret/terminator.rs:
```diff
@@ -174,8 +166,8 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
 
-            InlineAsm { template, ref operands, options, destination, .. } => {
-                M::eval_inline_asm(self, template, operands, options)?;
+            InlineAsm { template, operands, options, destination, .. } => {
```
There are now double references here where there weren't ones previously.

I don't really understand the point of these changes, this `match` expression in particular. Having to dereference values of types that are `Copy` seems like adding complexity to me, not removing it. To me, a lot of these changes seem like very slight improvements at best. :)

RalfJung replied:
Yeah I agree, it's better if the pattern can avoid creating `&Copy` references. Unfortunately due to https://github.com/rust-lang/rust/issues/64586 that often requires `ref` keywords, since match ergonimcs does not work well with `&` patterns. Still I think this is better than having `&Copy` references.

bors commented:
:umbrella: The latest upstream changes (presumably #106228) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

WaffleLapkin committed:
Don't call closures immediately, use `try{}` blocks

WaffleLapkin committed:
`rustc_hir_analysis`: some general code improvements

WaffleLapkin committed:
`rustc_hir_analysis`: remove `ref` patterns

WaffleLapkin committed:
tidy

rustfmt, pleaaaaase, start supporting rust

Co-authored-by: nils <48135649+Nilstrieb@users.noreply.github.com>

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3876061408/jobs/6609435453) [(plain)](https://github.com/rust-lang/rust/commit/958017aaed953e048475062fe9e890b9dab70558/checks/10531106329/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Successfully built 4fd6cfdd6c3c
Successfully tagged rust-ci:latest
Built container sha256:4fd6cfdd6c3c5862d9a3fb76ada230dcb9cca3950005c53f9ff1fcdd3c045361
Uploading finished image to https://ci-caches.rust-lang.org/docker/b8cdc76b9b2adf6aaac1172d56fab8a31470587d5348a0014d9e27d5bcbbe8577c6dad29501e9359634fe46a53fc6079832ca8161c7036f56b2737f782f779b6
upload failed: - to s3://rust-lang-ci-sccache2/docker/b8cdc76b9b2adf6aaac1172d56fab8a31470587d5348a0014d9e27d5bcbbe8577c6dad29501e9359634fe46a53fc6079832ca8161c7036f56b2737f782f779b6 Unable to locate credentials
[CI_JOB_NAME=mingw-check-tidy]
[CI_JOB_NAME=mingw-check-tidy]
---
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
---
tidy check
tidy: Skipping binary file check, read-only filesystem
Found 505 error codes
Highest error code: `E0791`
tidy error: /checkout/compiler/rustc_mir_build/src/thir/pattern/deconstruct_pat.rs:152: trailing whitespace
tidy error: /checkout/compiler/rustc_mir_build/src/build/expr/stmt.rs:116: trailing whitespace
some tidy checks failed
Build completed unsuccessfully in 0:00:22
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #106760) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

WaffleLapkin committed:
Undo questionable changes

WaffleLapkin commented:
I think I addressed all review comments. @rustbot ready

bors commented:
:umbrella: The latest upstream changes (presumably #106801) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Nilstrieb commented on compiler/rustc_hir_analysis/src/check/compare_impl_item.rs:
```diff
@@ -45,42 +45,22 @@ pub(super) fn compare_impl_method<'tcx>(
-    ) {
-        return;
-    }
+    let _: Result<_, ErrorGuaranteed> = try {
```
I appreciate all the small cleanups here, but could you make them in a separate PR next time?

WaffleLapkin replied:
sure, sorry 😅

Nilstrieb commented on compiler/rustc_hir_analysis/src/check/compare_impl_item.rs:
```diff
@@ -1242,41 +1211,41 @@ fn compare_number_of_method_arguments<'tcx>(
+            })
+            .or(trait_item_span);
+
+        let impl_span = {
+            let ImplItemKind::Fn(impl_m_sig, _) = &tcx.hir().expect_impl_item(impl_m.def_id.expect_local()).kind else { bug!("{:?} is not a method", impl_m) };
+            let pos = impl_number_args.saturating_sub(1);
+            impl_m_sig
```
```suggestion
        let ImplItemKind::Fn(impl_m_sig, _) = &tcx.hir().expect_impl_item(impl_m.def_id.expect_local()).kind else { bug!("{:?} is not a method", impl_m) };
        let pos = impl_number_args.saturating_sub(1);
        let impl_span = impl_m_sig
```
I don't see a point reducing the scope of the bindings here

WaffleLapkin replied:
I did this to not pollute the outer scope (e.g. `pos` is not very descriptive), but if you think that's not an issue I can remove the scope

Nilstrieb commented on compiler/rustc_hir_analysis/src/check/compare_impl_item.rs:
```diff
@@ -961,25 +938,18 @@ fn extract_spans_for_error_reporting<'tcx>(
-        }
-        _ => bug!("{:?} is not a method", impl_m),
+    let mut impl_args = {
+        let ImplItemKind::Fn(sig, _) = &tcx.hir().expect_impl_item(impl_m.def_id.expect_local()).kind else { bug!("{:?} is not a method", impl_m) };
```
All of these bugs make me think that we should have `expect_*` methods on `ImplItemKind` - but that's something for the future

Nilstrieb commented on compiler/rustc_trait_selection/src/traits/query/normalize.rs:
```diff
@@ -344,14 +344,14 @@ impl<'cx, 'tcx> FallibleTypeFolder<'tcx> for QueryNormalizer<'cx, 'tcx> {
             }
 
-            _ => ty.try_super_fold_with(self),
-        })()?;
```
Very weird, I hope there wasn't something weird going on, but the change looks correct🤨. Well, there are quite a few of those around, I guess it's just an artifact of refactorings and made sense at some point.

WaffleLapkin replied:
Yeah, when I was removing these I was triple checking in case I've missed something ><

compiler-errors commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 93f40f77ea04d0f59f10f0dda242814e4df5cc44 with merge 5cb9c9d58b4d1ad3f602047be2697d7569345deb...
<!-- homu: {"type":"TryBuildStarted","head_sha":"93f40f77ea04d0f59f10f0dda242814e4df5cc44","merge_sha":"5cb9c9d58b4d1ad3f602047be2697d7569345deb"} -->

compiler-errors commented on compiler/rustc_trait_selection/src/traits/query/normalize.rs:
```diff
@@ -230,7 +230,7 @@ impl<'cx, 'tcx> FallibleTypeFolder<'tcx> for QueryNormalizer<'cx, 'tcx> {
                             bug!(
                                 "infinite recursion generic_ty: {:#?}, substs: {:#?}, \
-                                 concrete_ty: {:#?}, ty: {:#?}",
+                             concrete_ty: {:#?}, ty: {:#?}",
```
whats up w this indentation

Noratrieb reviewed (commented):
A few more comments, but looking good overall

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3915060395/jobs/6694216798)
Build commit: 5cb9c9d58b4d1ad3f602047be2697d7569345deb (`5cb9c9d58b4d1ad3f602047be2697d7569345deb`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3915060395/jobs/6694216798"},"merge_sha":"5cb9c9d58b4d1ad3f602047be2697d7569345deb"} -->

rust-timer commented:
Queued 5cb9c9d58b4d1ad3f602047be2697d7569345deb with parent 0b90256ada21c6a81b4c18f2c7a23151ab5fc232, future [comparison URL](https://perf.rust-lang.org/compare.html?start=0b90256ada21c6a81b4c18f2c7a23151ab5fc232&end=5cb9c9d58b4d1ad3f602047be2697d7569345deb).
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (5cb9c9d58b4d1ad3f602047be2697d7569345deb): [comparison URL](https://perf.rust-lang.org/compare.html?start=0b90256ada21c6a81b4c18f2c7a23151ab5fc232&end=5cb9c9d58b4d1ad3f602047be2697d7569345deb&stat=instructions:u).

### Overall result: ❌ regressions - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=0b90256ada21c6a81b4c18f2c7a23151ab5fc232&end=5cb9c9d58b4d1ad3f602047be2697d7569345deb&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.0%  | [0.9%, 1.1%]   | 2     |
| Regressions ❌ <br /> (secondary)  | 2.0%  | [1.4%, 2.5%]   | 12    |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.2% | [-0.2%, -0.2%] | 1     |
| All ❌✅ (primary)                 | 1.0%  | [0.9%, 1.1%]   | 2     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=0b90256ada21c6a81b4c18f2c7a23151ab5fc232&end=5cb9c9d58b4d1ad3f602047be2697d7569345deb&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.5% | [-2.5%, -2.5%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -2.5% | [-2.5%, -2.5%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=0b90256ada21c6a81b4c18f2c7a23151ab5fc232&end=5cb9c9d58b4d1ad3f602047be2697d7569345deb&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

WaffleLapkin commented on compiler/rustc_const_eval/src/interpret/operand.rs:
```diff
@@ -363,11 +363,11 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         Ok(match src.as_mplace_or_imm() {
-            Left(ref mplace) => {
-                if let Some(val) = self.read_immediate_from_mplace_raw(mplace)? {
+            Left(mplace) => {
```
Thanks

WaffleLapkin committed:
Review suggestions

WaffleLapkin committed:
Self review suggestions

- add back accidentally removed new lines
- try to deref in patterns, rather than in expressions
  (maybe this was the reason of perf regression?...)

WaffleLapkin commented:
Here I am hoping that the regression is magically fixed.
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 258da9d27f286ac12da78b4df4957183d050db85 with merge 5cbaf3ebb74e0134fd910edbb70800baaed9f84c...
<!-- homu: {"type":"TryBuildStarted","head_sha":"258da9d27f286ac12da78b4df4957183d050db85","merge_sha":"5cbaf3ebb74e0134fd910edbb70800baaed9f84c"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3917376321/jobs/6697918861)
Build commit: 5cbaf3ebb74e0134fd910edbb70800baaed9f84c (`5cbaf3ebb74e0134fd910edbb70800baaed9f84c`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3917376321/jobs/6697918861"},"merge_sha":"5cbaf3ebb74e0134fd910edbb70800baaed9f84c"} -->

rust-timer commented:
Queued 5cbaf3ebb74e0134fd910edbb70800baaed9f84c with parent e9e09083674a58d361cd805877be55b5856d2806, future [comparison URL](https://perf.rust-lang.org/compare.html?start=e9e09083674a58d361cd805877be55b5856d2806&end=5cbaf3ebb74e0134fd910edbb70800baaed9f84c).
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (5cbaf3ebb74e0134fd910edbb70800baaed9f84c): [comparison URL](https://perf.rust-lang.org/compare.html?start=e9e09083674a58d361cd805877be55b5856d2806&end=5cbaf3ebb74e0134fd910edbb70800baaed9f84c&stat=instructions:u).

### Overall result: ❌ regressions - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=e9e09083674a58d361cd805877be55b5856d2806&end=5cbaf3ebb74e0134fd910edbb70800baaed9f84c&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.1%  | [1.0%, 1.2%]   | 2     |
| Regressions ❌ <br /> (secondary)  | 2.3%  | [2.0%, 2.6%]   | 6     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.2% | [-0.2%, -0.2%] | 1     |
| All ❌✅ (primary)                 | 1.1%  | [1.0%, 1.2%]   | 2     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=e9e09083674a58d361cd805877be55b5856d2806&end=5cbaf3ebb74e0134fd910edbb70800baaed9f84c&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.1% | [2.1%, 2.1%] | 1     |
| Regressions ❌ <br /> (secondary)  | 4.6% | [4.6%, 4.6%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 2.1% | [2.1%, 2.1%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=e9e09083674a58d361cd805877be55b5856d2806&end=5cbaf3ebb74e0134fd910edbb70800baaed9f84c&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 3.0% | [2.4%, 3.5%] | 6     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

WaffleLapkin commented:
I have no idea what causes slight perf regression here, should I split it into smaller PRs so that they can be benched separately?...

Nilstrieb commented:
Running the cachegrind command locally shows that the regression for `keccak` is pretty simple:
```
 529,322,559  ???:<rustc_data_structures::obligation_forest::ObligationForest<rustc_trait_selection::traits::fulfill::PendingPredicateObligation>>::process_obligations::<rustc_trait_selection::traits::fulfill::FulfillProcessor>
```
I would suggest reverting the changes around the obligation forest and fulfill

WaffleLapkin committed:
Untouch back perf sensetive code :sweat_smile:

WaffleLapkin commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit dbc6bf80ae93fb49f582b577ea00940b3701b35a with merge a06de2fabc5a3ebf738763c0a4db89153389cad0...
<!-- homu: {"type":"TryBuildStarted","head_sha":"dbc6bf80ae93fb49f582b577ea00940b3701b35a","merge_sha":"a06de2fabc5a3ebf738763c0a4db89153389cad0"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3918941222/jobs/6700438368)
Build commit: a06de2fabc5a3ebf738763c0a4db89153389cad0 (`a06de2fabc5a3ebf738763c0a4db89153389cad0`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3918941222/jobs/6700438368"},"merge_sha":"a06de2fabc5a3ebf738763c0a4db89153389cad0"} -->

rust-timer commented:
Queued a06de2fabc5a3ebf738763c0a4db89153389cad0 with parent 44a500c8c187b245638684748f54bd6ec67e0b25, future [comparison URL](https://perf.rust-lang.org/compare.html?start=44a500c8c187b245638684748f54bd6ec67e0b25&end=a06de2fabc5a3ebf738763c0a4db89153389cad0).
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (a06de2fabc5a3ebf738763c0a4db89153389cad0): [comparison URL](https://perf.rust-lang.org/compare.html?start=44a500c8c187b245638684748f54bd6ec67e0b25&end=a06de2fabc5a3ebf738763c0a4db89153389cad0&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=44a500c8c187b245638684748f54bd6ec67e0b25&end=a06de2fabc5a3ebf738763c0a4db89153389cad0&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 0.3%  | [0.3%, 0.4%]   | 2     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.2% | [-0.2%, -0.2%] | 2     |
| All ❌✅ (primary)                 | -     | -              | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=44a500c8c187b245638684748f54bd6ec67e0b25&end=a06de2fabc5a3ebf738763c0a4db89153389cad0&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 4.8%  | [4.8%, 4.8%]   | 1     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -3.3% | [-3.3%, -3.3%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=44a500c8c187b245638684748f54bd6ec67e0b25&end=a06de2fabc5a3ebf738763c0a4db89153389cad0&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

WaffleLapkin commented:
From what I can see the regressions look like noise, `deep-vector` mostly regressed in LLVM and `coercions` change seems to be in parsing, which I didn't change either. Is that ok?

Nilstrieb commented:
this does look like noise to me, so we're good to go!

Dear bors,

@bors r+

Love,
@Nilstrieb

bors commented:
:pushpin: Commit ed36f2e9cbee0bbdbe2a7b5bb7d65e4d102bd9c8 has been approved by `Nilstrieb`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Nilstrieb ed36f2e9cbee0bbdbe2a7b5bb7d65e4d102bd9c8 -->
<!-- homu: {"type":"Approved","sha":"ed36f2e9cbee0bbdbe2a7b5bb7d65e4d102bd9c8","approver":"Nilstrieb","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit ed36f2e9cbee0bbdbe2a7b5bb7d65e4d102bd9c8 with merge 65e3ee098ad6bf47e8fde1bbb043e4eb6357e6fb...
<!-- homu: {"type":"BuildStarted","head_sha":"ed36f2e9cbee0bbdbe2a7b5bb7d65e4d102bd9c8","merge_sha":"65e3ee098ad6bf47e8fde1bbb043e4eb6357e6fb"} -->

WaffleLapkin commented:
@bors retry
looks spurious (is github borked still?)

bors commented:
:hourglass: Testing commit ed36f2e9cbee0bbdbe2a7b5bb7d65e4d102bd9c8 with merge cc650431f5090f466fabb45c2b43a741a4e54f8d...
<!-- homu: {"type":"BuildStarted","head_sha":"ed36f2e9cbee0bbdbe2a7b5bb7d65e4d102bd9c8","merge_sha":"cc650431f5090f466fabb45c2b43a741a4e54f8d"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3934462471/jobs/6729449851)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/3934462471/jobs/6729449851","builder_name":"checks-actions"} -->

WaffleLapkin committed:
Hack compiletest to fix a random CI failure

WaffleLapkin commented:
@bors try
The CI failure should be fixed, but I can't test this locally.

WaffleLapkin commented:
The builder that was failing now passes, so I think this should be ready to be merged.

Nilstrieb commented on src/tools/compiletest/src/read2.rs:
```diff
@@ -110,9 +110,18 @@ impl ProcOutput {
-            ProcOutput::Abbreviated { mut head, skipped, tail } => {
+            ProcOutput::Abbreviated { mut head, mut skipped, tail } => {
+                let mut tail = &*tail;
+
+                // Skip over '{' at the start of the tail, so we don't later wrongfully consider this as json.
+                // See <https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/Weird.20CI.20failure/near/321797811>
+                while tail.get(0) == Some(&b'{') {
+                    tail = &tail[1..];
+                    skipped += 1;
+                }
```
this is, uh, certainly something. fun that no one has hit this before, you were the lucky one!

Nilstrieb commented:
That was quite a PR, as I said already it would be nice if it was better split-up next time. But anyways, should be good to go now!
@bors r+

bors commented:
:pushpin: Commit 65d1e8d9b556095dc50dc357e1ec4899afc3e975 has been approved by `Nilstrieb`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Nilstrieb 65d1e8d9b556095dc50dc357e1ec4899afc3e975 -->
<!-- homu: {"type":"Approved","sha":"65d1e8d9b556095dc50dc357e1ec4899afc3e975","approver":"Nilstrieb","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 65d1e8d9b556095dc50dc357e1ec4899afc3e975 with merge 56ee85274e5a3a4dda92f3bf73d1664c74ff9c15...
<!-- homu: {"type":"BuildStarted","head_sha":"65d1e8d9b556095dc50dc357e1ec4899afc3e975","merge_sha":"56ee85274e5a3a4dda92f3bf73d1664c74ff9c15"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3964842191/jobs/6795916500)
Approved by: Nilstrieb
Pushing 56ee85274e5a3a4dda92f3bf73d1664c74ff9c15 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Nilstrieb","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3964842191/jobs/6795916500"},"merge_sha":"56ee85274e5a3a4dda92f3bf73d1664c74ff9c15"} -->

rust-timer commented:
Finished benchmarking commit (56ee85274e5a3a4dda92f3bf73d1664c74ff9c15): [comparison URL](https://perf.rust-lang.org/compare.html?start=51d50ea96ecc9c681a0054e5eb8e5e1d4ab38906&end=56ee85274e5a3a4dda92f3bf73d1664c74ff9c15&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=51d50ea96ecc9c681a0054e5eb8e5e1d4ab38906&end=56ee85274e5a3a4dda92f3bf73d1664c74ff9c15&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.2% | [-0.2%, -0.1%] | 3     |
| Improvements ✅ <br /> (secondary) | -0.3% | [-0.5%, -0.3%] | 5     |
| All ❌✅ (primary)                 | -0.2% | [-0.2%, -0.1%] | 3     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=51d50ea96ecc9c681a0054e5eb8e5e1d4ab38906&end=56ee85274e5a3a4dda92f3bf73d1664c74ff9c15&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 2.1%  | [1.9%, 2.2%]   | 2     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -1.2% | [-1.2%, -1.2%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=51d50ea96ecc9c681a0054e5eb8e5e1d4ab38906&end=56ee85274e5a3a4dda92f3bf73d1664c74ff9c15&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -3.3% | [-3.3%, -3.3%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>
